src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -34,459 +34,682 @@
 }
 
-void ShenandoahSetsOfFree::move_to_set(size_t idx, ShenandoahFreeMemoryType new_set, size_t region_capacity) {
+void ShenandoahRegionPartitions::move_from_partition_to_partition(idx_t idx, ShenandoahFreeSetPartitionId orig_partition,
+                                                                  ShenandoahFreeSetPartitionId new_partition, size_t available) {
+  ShenandoahHeapRegion* r = ShenandoahHeap::heap()->get_region(idx);
   assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT, idx, _max);
-  assert ((new_set > NotFree) && (new_set < NumFreeSets), "New set must be valid");
-  ShenandoahFreeMemoryType orig_set = _membership[idx];
-  assert ((orig_set > NotFree) && (orig_set < NumFreeSets), "Cannot move free unless already free");
+  assert (orig_partition < NumPartitions, "Original partition must be valid");
+  assert (new_partition < NumPartitions, "New partition must be valid");
+  assert (available <= _region_size_bytes, "Available cannot exceed region size");
+  assert (_membership[int(orig_partition)].is_set(idx), "Cannot move from partition unless in partition");
+  assert ((r != nullptr) && ((r->is_trash() && (available == _region_size_bytes)) ||
+                             (r->used() + available == _region_size_bytes)),
+          "Used: " SIZE_FORMAT " + available: " SIZE_FORMAT " should equal region size: " SIZE_FORMAT,
+          ShenandoahHeap::heap()->get_region(idx)->used(), available, _region_size_bytes);
+
   // Expected transitions:
-  //  During rebuild: Mutator => Collector
-  //                  Mutator empty => Collector
-  //  During flip_to_gc:
-  //                  Mutator empty => Collector
-  //                  Mutator empty => Old Collector
-  // At start of update refs:
-  //                  Collector => Mutator
-  //                  OldCollector Empty => Mutator
-  assert((region_capacity <= _region_size_bytes && ((orig_set == Mutator && new_set == Collector) || (orig_set == Collector && new_set == Mutator)))
-      || (region_capacity == _region_size_bytes && ((orig_set == Mutator && new_set == Collector) || (orig_set == OldCollector && new_set == Mutator) || new_set == OldCollector)),
-      "Unexpected movement between sets");
-
-  _membership[idx] = new_set;
-  _capacity_of[orig_set] -= region_capacity;
-  shrink_bounds_if_touched(orig_set, idx);
-
-  _capacity_of[new_set] += region_capacity;
-  expand_bounds_maybe(new_set, idx, region_capacity);
-
-  _region_counts[orig_set]--;
-  _region_counts[new_set]++;
+  //  During rebuild:         Mutator => Collector
+  //                          Mutator empty => Collector
+  //                          Mutator empty => OldCollector
+  //  During flip_to_gc:      Mutator empty => Collector
+  //                          Mutator empty => OldCollector
+  // At start of update refs: Collector => Mutator
+  //                          OldCollector Empty => Mutator
+  assert (((available <= _region_size_bytes) &&
```

Comment:
This is quite hard for my feeble brain to parse. Could we break this into multiple asserts? or break some of the expressions out into functions?